src/support.rs

```diff
@@ -20,48 +18,64 @@ pub trait FromRawSurface {
     unsafe fn from_raw_surface(surface: *mut ffi::cairo_surface_t) -> Self;
 }
 
+#[derive(Debug)]
+struct CallbackEnv<W> {
+    writer: W,
+    error: Option<io::Error>,
+}
+
 #[derive(Debug)]
 pub struct Writer<S: FromRawSurface + AsRef<Surface>, W: io::Write> {
     pub surface: S,
-    writer: Box<W>,
+    callback_env: Box<CallbackEnv<W>>,
 }
 
 impl<S: FromRawSurface + AsRef<Surface>, W: io::Write> Writer<S, W> {
-    extern fn write_cb(writer: *mut c_void, data: *mut c_uchar, length: c_uint) -> cairo_status_t {
-        let mut writer: Box<W> = unsafe { Box::from_raw(writer as *mut _) };
+    extern fn write_cb(env: *mut c_void, data: *mut c_uchar, length: c_uint) -> cairo_status_t {
+        let env: &mut CallbackEnv<W> = unsafe { &mut *(env as *mut CallbackEnv<W>) };
```

Comment:
Should we check here if an error was already set before but for some reason we were called again, and in that case immediately error out?

Reply:
Good point!

I think both behaviors are reasonable, I don’t have a strong preference. Not attempting to write again after there has been an error makes sense. I wonder if cairo does the same after the callback returns `CAIRO_STATUS_WRITE_ERROR`